### PR TITLE
add parameter sync for batchnorm

### PR DESCRIPTION
**Pull request by wzhongyuan**
## What changes were proposed in this pull request?
1) Add parameter synchronizer
2) apply in batch normalization layers so that we can use small batch size to leverage more computing resources
## How was this patch tested?

Unit test, Resnet-50 training

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import scala.reflect._
+
+object ParameterSynchronizer {
+
+  val fEvents = new java.util.concurrent.ConcurrentHashMap[String, Event[Float]]
+
+  val dEvents = new java.util.concurrent.ConcurrentHashMap[String, Event[Double]]
+
+  def register[T: ClassTag](eventKey: String, threadNum: Int): Unit = {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: all public methods should have comments

**Comment by qiuxin2012** [comment]
What will happen if one of the executors is lost?

**Comment by wzhongyuan** [comment]
@qiuxin2012  it has nothing to do with executor lost, as the synchronization is inside executor , we don't do cross-executor sync

**Comment by wzhongyuan** [comment]
PR validation passed , http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/55/

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalizationSpec.scala [hunk 2]
```diff
@@ -17,13 +17,116 @@ package com.intel.analytics.bigdl.nn
...
+
+    val gradInput = bn.backward(input, gradOutput)
+
+    val saveMean = bn.saveMean
+    val saveStd = bn.saveStd
+    val runningMean = bn.runningMean
+    val runningVar = bn.runningVar
+
+    val bn1 = SpatialBatchNormalization[Float](2)
+    bn1.setParallism(Some(2))
```
[the remark this answers is not in the document]
Comment (wzhongyuan) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalizationSpec.scala]: Yes, it's an optimization, previously we just use Enginde,coreNumber by default, but it relies on Engine and does not make sense to expose to layer itselft , also it's not flexible,

**Comment by wzhongyuan** [comment]
@qiuxin2012  Please let me know if you have other concerns, thanks

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 3]
```diff
@@ -61,6 +61,19 @@ class BatchNormalization[T: ClassTag](
 
   require(nOutput > 0, "output feature map number must be greater than zero")
 
+  private var parallism : Option[Int] = None
+
+  def setParallism(parallism: Option[Int]): Unit = {
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: comments for public methods and variables

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: Maybe
```scala
def setParallism(parallism: Int): Unit = {
    this.parallism = Some(parallism)
}
```
is more user friendly. And should it be private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 4]
```diff
@@ -209,21 +246,37 @@ class BatchNormalization[T: ClassTag](
     saveMean.resizeAs(runningMean).zero
     saveStd.resizeAs(runningVar).fill(ev.zero)
 
+    val nChannels = _input.size(2)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: should we requirement the format is NCHW?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 5]
```diff
@@ -243,25 +296,39 @@ class BatchNormalization[T: ClassTag](
...
+    val gmKeyWithId = s"${this.gmKey}_${this.getId}"
+    val gxmKeyWithId = s"${this.gxmKey}_${this.getId}"
+    val needSync = getParallism() != None && getParallism().get > 1
     _gradOutput.set(gradOutput)
     makeBatch(_gradOutput)
     _gradOutput.addSingletonDimension(_gradOutput, 3)
     _gradOutput.addSingletonDimension(_gradOutput, 4)
     gxMean.zero()
     gMean.zero()
+    val nChannel = _gradOutput.size(2)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: should we require the input is NCHW?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import org.apache.spark.TaskContext
+
+import scala.reflect._
+
+/**
+ * A parameter synchronizer among threads per task
+ */
+object ParameterSynchronizer {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: private[bigdl] ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+  val fEvents = new java.util.concurrent.ConcurrentHashMap[String, Event[Float]]
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: why there're two kind of events?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      if (event == null) {
+        event = new Event[Float](threadNum)
+        fEvents.putIfAbsent(eventKey, event)
+      }
+    } else if (classTag[T] ==  classTag[Double]) {
+      var event = dEvents.get(eventKey)
+      if (event == null) {
+        event = new Event[Double](threadNum)
+        dEvents.putIfAbsent(eventKey, event)
+      }
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: else throw exception?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      fEvents.get(eventKey).data.asInstanceOf[java.util.Map[String, Tensor[T]]]
+    } else if (classTag[T] ==  classTag[Double]) {
+      dEvents.get(eventKey).data.asInstanceOf[java.util.Map[String, Tensor[T]]]
+    } else {
+      null
+    }
+  }
+}
+
+class Event[T: ClassTag](threadNum: Int) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: private?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala (hunk 4 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: i's already separated in NCHW related methods, not impact on others

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala (hunk 5 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: it's already separated in NCHW related methods, not impact on others

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala (hunk 8 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: we cannot throw exception here ,  it's for race condition and make sure the thread safety

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala (hunk 7 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/ParameterSynchronizer.scala]: For different types, as `SpatialBatchNormalization` implements explicitly type specific methods, this way would reduce the implicit conversion cost

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala (hunk 3 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: Makes sense, we cannot make it private , it's an exposed interface

**Comment by wzhongyuan** [comment]
@yiheng ,  Ran another PR job and passed after refinement http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/61/ 

Please let me know if any other issue

**Comment by wzhongyuan** [comment]
Per review refinement PR validation passed 
http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/65/
